src/python/pants/core/goals/tailor.py

```diff
@@ -117,9 +121,11 @@ def for_target_type(
         comments: Iterable[str] = tuple(),
         build_file_name: str = "BUILD",
     ):
-        owned_sources = (
-            (kwargs or {}).get("sources") or default_sources_for_target_type(target_type) or tuple()
-        )
+        maybe_owned_sources = (kwargs or {}).get("sources")
+        default_sources = default_sources_for_target_type(target_type)
+        if (maybe_owned_sources or triggering_sources) and default_sources is None:
+            raise ValueError(f"Target type {target_type.__name__} does not have a sources field.")
```

Comment:
Let's be even more verbose here: "A target of type {target_type.__name__} was proposed at address {path}:{name} with explicit sources {','.join(explicit_sources or triggering sources)}, but this target type does not have a sources field."

Reply:
Done. I also added a test for this case and fixed a bug in checking `default_sources` where it is an empty tuple and not `None` thus the `None` check is insufficient to check if it is empty. Thus switched to a false-y check. (`not default_sources`).